docs/8-kubernetes-container-orchestration/8.7.1-kustomize.md

```diff
@@ -0,0 +1,277 @@
+
+> Also note that some fields are immutable in Kubernetes. You may need to delete your resources and reapply. Delete by running `kustomize build . | kubectl delete -f -`
+
+## Exercise 4: Secret Management
+
+Often times, you need to create/apply secrets into your Kubernetes cluster so different containers can access them. While under the right circumstances, it's okay for the secrets to be visible locally and on your cluster (ie not completely hidden), it is never okay for them to be pushed to GitHub. Kustomize can help you structure your resources in a way that the secret *existence* will be visible in GitHub, but not the secret content itself.
+
+* Start off by creating a `.env` file and place a key,value pair inside of it. This file will only be stored locally, and should never be pushed to GitHub.
+
+`.env`
+```bash
+SECRET=123456
+```
+
+`.gitignore`
+```bash
+.env
+```
+
+* Now that your secret information is stored locally, look into the Kustomize `secretGenerator` to, well, generate a secret that contains your key,value pair.
+> Note that the secret will be created with base64 encoded information, so you'll need to decode it to actually read the contents of the secret.
+
+* Finally, write a Kustomize `patch` (either method from above) to add the following entry to your container's `env` list in the deployment:
+
+```yaml
+- name: SECRET
+  valueFrom:
+    secretKeyRef:
+      name: <The name you gave your secret>
+      key: SECRET
+```
+
+* Exec into your container running in the cluster and run `env` to confirm that your environment variable is present and holding the correct information.
+> Since you cannot change a containers base env list after booting, you may need to delete your deployment before applying this change.
+
+* Now that your secret exists, update your key,value pair in your `.env` file and reapply. Did your secret get updated? If so, good. If it didn't and instead a *new* secret was created with your change, figure out how to correct that.
+
+## Exercise 5: Script Embedding
+
+Earlier in the chapter you were shown the `ConfigMap` resource. As a refresher, a `ConfigMap` is a resource that holds *unstructured* data (basically whatever you want, stored as a string). ConfigMaps are commonly used to embed scripts into Kubernetes clusters, and are then mounted to containers for execution. However, the process of writing these scripts is painful, because you need to write them as one long string inside of a ConfigMap field.
+
+Yep, you guessed it. Kustomize solves this problem. For this exercise, we're going to leverage Kustomize to overwrite the frontend for the Nginx server.
+
+* Start by creating an `index.html` file, and include some HTML code similar to the following:
+```html
+<html>
+    <h1>Apprentices Are Real People Too</h1>
+    <body>
+        This is my super cool Nginx Server
+    </body>
+</html>
```

Comment:
```suggestion
<html>
    <body>
        <h1>Apprentices Are Real People Too</h1>
        This is my super cool Nginx Server
    </body>
</html>
```

All visible DOM elements per the HTML specifications should be in the `<body>` tags